### PR TITLE
Support for vfat filesystem

### PR DESCRIPTION
**Pull request by vojtechtrefny**

### CODE REVIEW

File: src/plugins/fs.c [hunk 1]
```diff
@@ -267,7 +315,77 @@ gboolean bd_fs_wipe (const gchar *device, gboolean all, GError **error) {
...
+        blkid_free_probe (probe);
+        synced_close (fd);
+        return FALSE;
+    }
+
+	blkid_probe_enable_partitions(probe, 1);
+	blkid_probe_set_partitions_flags(probe, BLKID_PARTS_MAGIC);
+	blkid_probe_enable_superblocks(probe, 1);
+	blkid_probe_set_superblocks_flags(probe, BLKID_SUBLKS_USAGE | BLKID_SUBLKS_TYPE |
+                                               BLKID_SUBLKS_MAGIC | BLKID_SUBLKS_BADCSUM);
```
Comment (vpodzime) [inline src/plugins/fs.c]: These lines have some weird indentation.

File: src/plugins/fs.c [hunk 2]
```diff
@@ -373,6 +491,33 @@ static gboolean wipe_fs (const gchar *device, const gchar *fs_type, GError **err
         return FALSE;
     }
 
+    if (wipe_all) {
+        gboolean has_fat = TRUE;
+        has_fat = has_fs (device, "vfat", error);
```
Comment (vpodzime) [inline src/plugins/fs.c]: This cannot be vfat-specific, it needs to take the ``fs_type`` parameter into account.

Comment (vpodzime) [inline src/plugins/fs.c]: And it looks like opening and probing the device over and over by calling this function is not necessary when the ``wipe_fs`` function has the device opened and is probing it.

File: src/plugins/fs.c [hunk 3]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+ * bd_fs_vfat_check:
+ * @device: the device containing the file system to check
+ * @extra: (allow-none) (array zero-terminated=1): extra options for the repair (right now
+ *                                                 passed to the 'fsck.vfat' utility)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether an vfat file system on the @device is clean or not
+ *
+ * Note: if the file system is mounted it may be reported as unclean even if
+ *       everything is okay and there are just some pending/in-progress writes
```
Comment (vpodzime) [inline src/plugins/fs.c]: Is this really true for *vfat* or is it just copied over from the ``ext4_check`` function?

File: src/plugins/fs.c [hunk 4]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+gboolean bd_fs_vfat_resize (const gchar *device, guint64 new_size, GError **error) {
+    PedDevice *ped_dev = NULL;
+    PedGeometry geom = {0};
+    PedGeometry new_geom = {0};
+    PedFileSystem *fs = NULL;
+    PedSector start = 0;
+    PedSector length = 0;
+    gint status = 0;
+
+    ped_dev = ped_device_get(device);
```
Comment (vpodzime) [inline src/plugins/fs.c]: Please add spaces before parentheses in function calls.

File: src/plugins/fs.c [hunk 5]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+    status = blkid_probe_set_device (probe, fd, 0, 0);
+    if (status != 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to create a probe for the device '%s'", device);
+        blkid_free_probe (probe);
+        synced_close (fd);
+        return FALSE;
+    }
+
+	blkid_probe_enable_partitions(probe, 1);
```
Comment (vpodzime) [inline src/plugins/fs.c]: weird indentation

File: src/plugins/fs.c [hunk 6]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+    PedFileSystem *fs = NULL;
+    PedSector start = 0;
+    PedSector length = 0;
+    gint status = 0;
+
+    ped_dev = ped_device_get(device);
+    if (!ped_dev) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to get ped device for the device '%s'", device);
+      return FALSE;
```
Comment (vpodzime) [inline src/plugins/fs.c]: weird indentation

File: src/plugins/fs.c [hunk 7]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+    }
+
+    status = ped_device_open(ped_dev);
+    if (status == 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to get open the device '%s'", device);
+      return FALSE;
```
Comment (vpodzime) [inline src/plugins/fs.c]: same here

File: src/plugins/part_err.c [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+#include "part_err.h"
+
+static __thread gchar *error_msg = NULL;
+
+PedExceptionOption exc_handler (PedException *ex) {
+    error_msg = g_strdup (ex->message);
+    return PED_EXCEPTION_UNHANDLED;
+}
+
+gchar * get_error_msg () {
```
Comment (vpodzime) [inline src/plugins/part_err.c]: Please prefix names of both these functions with the library's prefix.

File: src/plugins/fs.c [hunk 9]
```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+/**
+ * bd_fs_vfat_resize:
+ * @device: the device the file system of which to resize
+ * @new_size: new requested size for the file system (if 0, the file system is
+ *            adapted to the underlying block device)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the file system on @device was successfully resized or not
+ */
+gboolean bd_fs_vfat_resize (const gchar *device, guint64 new_size, GError **error) {
```
Comment (vpodzime) [inline src/plugins/fs.c]: This function needs to take care about progress reporting on it's own as it doesn't run any utility.

File: src/plugins/part_err.h [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#include <glib.h>
+#include <parted/parted.h>
+
+#include <utils.h>
+#include <fs.h>
+#include <part.h>
```
Comment (vpodzime) [inline src/plugins/part_err.h]: Why these three includes?

File: src/utils/exec.c [hunk 11]
```diff
@@ -214,6 +214,7 @@ gboolean bd_utils_exec_and_report_status_error (const gchar **argv, const BDExtr
     success = g_spawn_sync (NULL, args ? (gchar **) args : (gchar **) argv, NULL, G_SPAWN_SEARCH_PATH,
                             (GSpawnChildSetupFunc) set_c_locale, NULL,
                             &stdout_data, &stderr_data, status, error);
+    *status = *status % 255;
```
Comment (vpodzime) [inline src/utils/exec.c]: I think this is worth a comment.

**Comment by vojtechtrefny** [comment]
New version based on @vpodzime comments. Test failures are not related to this changes.

File: src/plugins/fs.c [hunk 12]
```diff
@@ -373,6 +451,38 @@ static gboolean wipe_fs (const gchar *device, const gchar *fs_type, GError **err
         return FALSE;
     }
 
+    blkid_reset_probe (probe);
+
+    if (wipe_all) {
+        gboolean has_fs_type = TRUE;
```
Comment (vpodzime) [inline src/plugins/fs.c]: Please move the variable declaration+definition to the start of the function.

File: src/plugins/fs.c [hunk 13]
```diff
@@ -373,6 +451,38 @@ static gboolean wipe_fs (const gchar *device, const gchar *fs_type, GError **err
...
+        has_fs_type = has_fs (probe, device, fs_type, error);
+
+        while (has_fs_type) {
+            status = blkid_do_probe (probe);
+            if (status != 0) {
+                g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                             "Failed to probe the device '%s'", device);
+                blkid_free_probe (probe);
+                synced_close (fd);
```
Comment (vpodzime) [inline src/plugins/fs.c]: Report finish (failure) of the action here.

File: src/plugins/fs.c [hunk 14]
```diff
@@ -817,3 +927,324 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
...
+        g_set_error (error, BD_FS_ERROR, type,
+                     " (%s)", error_msg);
+        g_free (error_msg);
+        error_msg = NULL;
+        return TRUE;
+    } else {
+        g_set_error_literal (error, BD_FS_ERROR, type, "");
+        return FALSE;
+    }
+}
```
Comment (vpodzime) [inline src/plugins/fs.c]: Please move these two functions near the top of the file.

**Comment by vojtechtrefny** [comment]
Fixed.

**Pull request merged**